### ISSUE TITLE
"Canvas" created with "NumSamples" different to "One" display a stretched, flipped image

### ISSUE THREAD

**Issue by Athorus** [ggez/ggez#695]
**Describe the bug**
When calling Canvas::new( ) with NumSamples::One it works as expected. But if you call it with a different value, a black screen is displayed.

**To Reproduce**
See code below.

**Expected behavior**
Any variant of NumSamples should work. Only the AA filter should change.

**Screenshots or pasted code**
```rust
use ggez::ContextBuilder;
use ggez::conf::{Conf, NumSamples};
use ggez::Context;
use ggez::event::{self, EventHandler};
use ggez::GameResult;
use ggez::graphics::{self, Canvas, Color, DrawParam, Image};

struct Application {
    canvas: Canvas
}

impl Application {
    fn new(ctx: &mut Context) -> Self {
        let image = Image::new(ctx, "/light-image.png").unwrap();
        let canvas = Self::canvas(ctx, &image);

        Self {canvas}
    }

    fn canvas(ctx: &mut Context, image: &Image) -> Canvas {
        let (w, h) = graphics::drawable_size(ctx);
        //let canvas = Canvas::new(ctx, w as u16, h as u16, NumSamples::One).unwrap();
        let canvas = Canvas::new(ctx, w as u16, h as u16, NumSamples::Four).unwrap();
        graphics::set_canvas(ctx, Some(&canvas));
        graphics::draw(ctx, image, DrawParam::new()).unwrap();
        graphics::set_canvas(ctx, None);

        canvas
    }
}

impl EventHandler for Application {
    fn update(&mut self, _: &mut Context) -> GameResult<()> {
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        graphics::clear(ctx, Color{r: 0., g: 0.3, b: 0., a: 1.});
   
        graphics::draw(ctx, &self.canvas, DrawParam::new()).unwrap();

        graphics::present(ctx).unwrap();
        Ok(())
    }
}

pub fn main() {
    let c = Conf::new();
    let (ctx, event_loop) = &mut ContextBuilder::new("canvas", "athorus").conf(c).build().unwrap();
    let app = &mut Application::new(ctx);
    event::run(ctx, event_loop, app).unwrap();
}
```

**Hardware and Software:**
 - ggez version: 0.5.1
 - OS: Windows 10
 - Graphics card: AMD Radeon(TM) R9 290X
 - Graphics card drivers: Radeon 19.9.2

**Comment by icefoxen** [comment]
Multisampling in canvases is... shall we say, not recommended, at least right now.  https://github.com/ggez/ggez/issues/411 might be related.

**Comment by Athorus** [comment]
Thanks for your answers.

I need Canvas, because I draw the backgroud of my game with a very complex Mesh. In order to increase FPS, I use a Canvas (drawing the mesh only once).
That is why I need multisampling too.

**Comment by Athorus** [comment]
Well, I've done some test and the speed difference is small. This is surprising how drawing a Mesh is fast compared to drawing a Canvas.

**Comment by icefoxen** [comment]
From what I understand about the process, I'd recommend only doing multisampling once anywhere in your pipeline, and the simplest place to put it is in the last step where things are actually drawn to the screen's buffer.  That's what I usually do, so I haven't run across this before, but either way ggez's API offers the capability so we should make it actually work.  Looks like there might be an extra step or two in `gfx` that needs to be set, or it might be the hardware being reluctant to support it; I'll have to try it out myself and see.

Drawing a  small-ish Mesh ("small-ish" meaning 1000's of triangles) probably doesn't come close to saturating the abilities of most modern GPU's, so I'm not super surprised.  If you end up drawing a lot of *other* stuff as well then the difference may start mattering more.

**Comment by Athorus** [comment]
I'm not sure to understand everything. Thanks to ggez 0.5.1, is it possible to do "multisampling once anywhere in your pipeline, and the simplest place to put it is in the last step where things are actually drawn to the screen's buffer." ?
Currently, I build a "Context" with a "WindowSetup" containing "NumSamples::Four", but multisampling is not applied when I draw "Canvas".

**Comment by nobbele** [comment]
not related to #411

**Comment by PSteinhaus** [comment]
Ok, so this is still broken, but I found out that the example code here needs to be fixed a little, because ```graphics::present``` "unsets" all canvases, thereby creating the black image due to an empty canvas. But if we change it to this:
```rust
use ggez::ContextBuilder;
use ggez::conf::{Conf, NumSamples};
use ggez::Context;
use ggez::event::{self, EventHandler};
use ggez::GameResult;
use ggez::graphics::{self, Canvas, Color, DrawParam, Image, get_window_color_format};

struct Application {
    canvas: Canvas,
    image: Option<Image>
}

impl Application {
    fn new(ctx: &mut Context) -> Self {
        let image = Image::new(ctx, "/tile.png").unwrap();
        let canvas = Self::canvas(ctx, &image);

        Self {canvas, image: Some(image)}
    }

    fn canvas(ctx: &mut Context, image: &Image) -> Canvas {
        let (w, h) = graphics::drawable_size(ctx);
        //let canvas = Canvas::new(ctx, w as u16, h as u16, NumSamples::One).unwrap();
        let canvas = Canvas::new(ctx, w as u16, h as u16, NumSamples::Two, get_window_color_format(ctx)).unwrap();

        canvas
    }
}

impl EventHandler for Application {
    fn update(&mut self, _: &mut Context) -> GameResult<()> {
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        graphics::clear(ctx, Color{r: 0., g: 0.3, b: 0., a: 1.});

        graphics::set_canvas(ctx, Some(&self.canvas));
        graphics::draw(ctx, self.image.as_ref().unwrap(), DrawParam::new()).unwrap();
        graphics::set_canvas(ctx, None);

        graphics::draw(ctx, &self.canvas, DrawParam::new()).unwrap();

        graphics::present(ctx).unwrap();
        Ok(())
    }
}

pub fn main() {
    let (mut ctx, event_loop) = ContextBuilder::new("canvas", "athorus").build().unwrap();
    let app = Application::new(&mut ctx);
    event::run(ctx, event_loop, app)
}
```
then the behavior changes:
- with ```NumSamples::One``` we get what we expect, an image drawn in the top left
- with ```NumSamples::Two``` (or any other number except 1) we DON'T get a black screen; instead we see the image, but if fills the whole canvas and thereby the whole screen.

**Comment by PSteinhaus** [comment]
@icefoxen I suggest changing that last sentence in the documentation of ```graphics::present()``` from "Unsets any active canvas." to "Clears all active canvases." Or would that be inaccurate (i.e. does "unset" mean something else as well)?

**Comment by nobbele** [comment]
Why in the world would adding anti-aliasing stretch and FLIP the canvas

**Comment by PSteinhaus** [comment]
Canvases and strange flipping behavior really seem to belong together.

The only thing that makes sense for me right now would be that drawing the image on the anti-aliased canvas (somehow) just sets the texture of this canvas to be the image (since DrawParams on this call don't seem to have any effect, I just checked).

```Canvas::draw``` also (somehow) circumvents all the clever flipping stuff you just invented in #936 (since DrawParams on this call don't seem to have any effect either, I checked).

The only way I can explain this to myself is that both fail because ```draw_image_raw```, which both calls use, fails in the same way both times.

But _how_ it fails (i.e. creates this strange behavior) I don't know.

**Comment by PSteinhaus** [comment]
The workaround has been implemented in #991.

Now, a special fragment shader serves as our internal tool for resolving the multi-sampled texture into a regular non-multi-sampled one. Both of them are stored inside `Canvas`. The first is used as a target for drawing onto (with MSAA). The second is used when drawing the `Canvas` onto something else, or when it's serialized via `to_rgba8` (or a different function that calls it, such as `to_image`).

The link between both, resolving the first into the second one, is `Canvas::resolve`, which needs to be called before the then updated second texture (the non-multi-sampled one) can be used. Luckily ggez does this automatically, by calling `resolve` when calling `Canvas::draw` or `Canvas::to_rgba8`, so the end user doesn't need to think about this, unless they desire to manipulate the _raw_ image (which I chose to be the non-multi-sampled texture). If they want to work with that directly, then they need to call `resolve` to update the texture before working with it, but this is really a very specific edge case.

The only sad thing is that this workaround doesn't work for GLES, as the shader language doesn't include `sampler2DMS`, which is what the workaround is based on. Maybe there's a hacky way to write a fitting equivalent shader for GLES as well, but for now I've no idea what that could look like, or whether it's even possible.
